java/src/jmri/jmrit/logixng/actions/ActionBlock.java

```diff
@@ -0,0 +1,608 @@
+
+        switch (_operationAddressing) {
+            case Reference:
+                return ReferenceUtil.getReference(
+                        getConditionalNG().getSymbolTable(), _operationReference);
+
+            case LocalVariable:
+                SymbolTable symbolTable = getConditionalNG().getSymbolTable();
+                return TypeConversionUtil
+                        .convertToString(symbolTable.getValue(_operationLocalVariable), false);
+
+            case Formula:
+                return _operationExpressionNode != null
+                        ? TypeConversionUtil.convertToString(
+                                _operationExpressionNode.calculate(
+                                        getConditionalNG().getSymbolTable()), false)
+                        : null;
+
+            default:
+                throw new IllegalArgumentException("invalid _addressing state: " + _operationAddressing.name());
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void execute() throws JmriException {
+        Block block;
+
+        switch (_addressing) {
+            case Direct:
+                block = _blockHandle != null ? _blockHandle.getBean() : null;
+                break;
+
+            case Reference:
+                String ref = ReferenceUtil.getReference(
+                        getConditionalNG().getSymbolTable(), _reference);
+                block = InstanceManager.getDefault(jmri.BlockManager.class)
+                        .getNamedBean(ref);
+                break;
+
+            case LocalVariable:
+                SymbolTable symbolTable = getConditionalNG().getSymbolTable();
+                block = InstanceManager.getDefault(BlockManager.class)
+                        .getNamedBean(TypeConversionUtil
+                                .convertToString(symbolTable.getValue(_localVariable), false));
+                break;
+
+            case Formula:
+                block = _expressionNode != null ?
+                        InstanceManager.getDefault(BlockManager.class)
+                                .getNamedBean(TypeConversionUtil
+                                        .convertToString(_expressionNode.calculate(
+                                                getConditionalNG().getSymbolTable()), false))
+                        : null;
+                break;
+
+            default:
+                throw new IllegalArgumentException("invalid _addressing state: " + _addressing.name());
+        }
+
+        if (block == null) {
+            return;
+        }
+
+        String name = (_operationAddressing != NamedBeanAddressing.Direct)
+                ? getNewLock() : null;
+
+        if (_operationDirect != DirectOperation.None) {
+            DirectOperation oper;
+            if ((_operationAddressing == NamedBeanAddressing.Direct)) {
+                oper = _operationDirect;
+            } else {
+                oper = DirectOperation.valueOf(name);
+            }
```

Comment:
Not sure, but it looks like this code should be:

```
        String name = (_operationAddressing != NamedBeanAddressing.Direct)
                ? getNewLock() : null;

        DirectOperation oper;
        if ((_operationAddressing == NamedBeanAddressing.Direct)) {
            oper = _operationDirect;
        } else {
            oper = DirectOperation.valueOf(name);
        }

        if (oper != DirectOperation.None) {
```

Reply:
If the **lock** related code is left over from turnout locking, a lot of that can go away.

Reply:
No, it has nothing to do with locking. This code is for indirect addressing. In Logix, you can have the turnout `@M1`, which means that the turnout name is in the memory M1.

In LogixNG you have more flexibility. For example you can have a formula `"IT"+index` where `index` is a local variable. This might be useful if you want to set IT1, IT2, IT3, ..., IT10 to thrown. You can then have a For loop iterate index from 1 to 10.